### PR TITLE
Improve `TypeMapping`

### PR DESCRIPTION
**Pull request by emilyaherbert**
This PR cuts down on some of the unnecessary bloat of the `matches_type_parameter` function and replaces it with an API on an improved `TypeMapping` type.

### CODE REVIEW

File: sway-core/src/type_system/type_mapping.rs [hunk 1]
```diff
@@ -1,130 +1,253 @@
...
+                    insert_type(TypeInfo::UnknownGeneric {
+                        name: x.name_ident.clone(),
+                    }),
+                )
+            })
+            .collect();
+        TypeMapping { mapping }
+    }
+
+    pub(crate) fn from_superset_and_subset(
```
Comment (mohammadfawaz) [inline sway-core/src/type_system/type_mapping.rs:26]: Do you mind adding a comment before this method explaining what it does? Specifically, I'm looking for what `superset` and `subset` exactly refer to.

**Review by mohammadfawaz — COMMENTED** [review]
All looks good to me. It would be nice to see detailed documentation for all the methods in `TypeMapping` 🙌

File: sway-core/src/type_system/type_mapping.rs (hunk 1 above)
Comment (emilyaherbert) [reply sway-core/src/type_system/type_mapping.rs]: Done!

**Review by mohammadfawaz — APPROVED** [review]
Great! Thanks for all the additional comments :)

**Review by otrho — APPROVED** [review]

**Pull request merged**